File: public/packages/typescript/resilience/src/pipeline/singleflight-layer.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import { Singleflight } from "../singleflight/singleflight.js";
import { raceAbort } from "./abort.js";
import type { IResilientLayer } from "./i-resilient-layer.js";

/**
 * Layer wrapping an inner layer with key-based singleflight dedup. Mirrors
 * .NET `SingleflightLayer<TKey, TValue>` + `Singleflight.ExecuteAsync`.
 *
 * Per-caller cancellation affects ONLY that caller's wait, never the shared
 * operation: the deduplicated inner op runs with NO signal (`undefined`, the JS
 * analogue of `CancellationToken.None`), so one caller aborting cannot cancel
 * the shared work that the other waiters still depend on. Each caller's wait is
 * raced against that caller's own signal — an aborting caller's promise rejects
 * with `AbortError` while the shared op continues and the remaining waiters
 * receive its result.
 */
export class SingleflightLayer implements IResilientLayer {
  private readonly sf = new Singleflight<string, unknown>();

  constructor(private readonly inner: IResilientLayer) {}

  execute<T>(
    key: string,
    op: (signal?: AbortSignal) => Promise<T>,
    signal?: AbortSignal,
  ): Promise<T> {
    // The shared run gets NO signal — one caller's abort must not cancel it.
    const shared = this.sf.do(key, () =>
      this.inner.execute(key, op, undefined),
    ) as Promise<T>;
    // Each caller's wait IS cancellable by that caller's signal.
    return raceAbort(shared, signal);
  }
}

File: public/packages/dotnet/resilience/Pipeline/SingleflightLayer.cs
// -----------------------------------------------------------------------
// <copyright file="SingleflightLayer.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Resilience.Pipeline;

using DcsvIo.D2.Resilience.Singleflight;

/// <summary>
/// Pipeline layer that dedupes concurrent in-flight operations by the
/// per-call key via the wrapped <see cref="Singleflight{TKey, TValue}"/>
/// instance.
/// </summary>
/// <typeparam name="TKey">The deduplication key type.</typeparam>
/// <typeparam name="TValue">The value type produced by the operation.</typeparam>
public sealed class SingleflightLayer<TKey, TValue> : IResilientLayer<TKey, TValue>
    where TKey : notnull
{
    private readonly Singleflight<TKey, TValue> r_singleflight;

    /// <summary>
    /// Initializes a <see cref="SingleflightLayer{TKey, TValue}"/> wrapping
    /// the supplied <paramref name="singleflight"/> instance.
    /// </summary>
    /// <param name="singleflight">The Singleflight to wrap.</param>
    public SingleflightLayer(Singleflight<TKey, TValue> singleflight)
        => r_singleflight = singleflight;

    /// <inheritdoc/>
    public ValueTask<TValue> WrapAsync(
        TKey key,
        Func<CancellationToken, ValueTask<TValue>> next,
        CancellationToken ct)
        => r_singleflight.ExecuteAsync(key, next, ct);
}
